SYCL/Basic/reqd_work_group_size.cpp

```diff
@@ -12,6 +12,167 @@
+  });
+}
+
+template <TestKind Kind, int ReqdWGSizeX>
+void submitTest(queue &Q, range<1> GlobalRange, range<1> LocalRange) {
+  Q.submit([&](handler &CGH) {
+    CGH.parallel_for<ReqdWGSizeVariadic<Kind, ReqdWGSizeX>>(
+        nd_range<1>(GlobalRange, LocalRange),
+        [=](nd_item<1>) [[sycl::reqd_work_group_size(ReqdWGSizeX)]] {});
+  });
+}
+
+template <TestKind Kind, int ReqdWGSizeX, int ReqdWGSizeY, int ReqdWGSizeZ>
+void submitTest(queue &Q, range<3> GlobalRange) {
+  Q.submit([&](handler &CGH) {
+    CGH.parallel_for<
+        ReqdWGSizeVariadic<Kind, ReqdWGSizeX, ReqdWGSizeY, ReqdWGSizeZ>>(
+        GlobalRange, [=](item<3>) [[sycl::reqd_work_group_size(
+                         ReqdWGSizeX, ReqdWGSizeY, ReqdWGSizeZ)]] {});
+  });
+}
+
+template <TestKind Kind, int ReqdWGSizeX, int ReqdWGSizeY>
+void submitTest(queue &Q, range<2> GlobalRange) {
+  Q.submit([&](handler &CGH) {
+    CGH.parallel_for<ReqdWGSizeVariadic<Kind, ReqdWGSizeX, ReqdWGSizeY>>(
+        GlobalRange,
+        [=](item<2>)
+            [[sycl::reqd_work_group_size(ReqdWGSizeX, ReqdWGSizeY)]] {});
+  });
+}
+
+template <TestKind Kind, int ReqdWGSizeX>
+void submitTest(queue &Q, range<1> GlobalRange) {
+  Q.submit([&](handler &CGH) {
+    CGH.parallel_for<ReqdWGSizeVariadic<Kind, ReqdWGSizeX>>(
+        GlobalRange,
+        [=](item<1>) [[sycl::reqd_work_group_size(ReqdWGSizeX)]] {});
+  });
+}
+
+template <int... ReqdWGSizes>
+int runPositiveTest(queue &Q, range<sizeof...(ReqdWGSizes)> GlobalRange,
+                    range<sizeof...(ReqdWGSizes)> LocalRange) {
+  std::string TestCaseName =
+      "Test case ReqdWGSizeVariadic " +
+      GetRangeStr(range<sizeof...(ReqdWGSizes)>{ReqdWGSizes...}) + " & " +
+      GetRangeStr(LocalRange) + " (Positive with local range)";
+  try {
+    submitTest<TestKind::PositiveWithLocal, ReqdWGSizes...>(Q, GlobalRange,
+                                                            LocalRange);
+    Q.wait_and_throw();
+    return 0;
+  } catch (nd_range_error &E) {
+    std::cerr << TestCaseName
+              << " failed: unexpected nd_range_error exception: " << E.what()
+              << std::endl;
+    return 1;
+  } catch (runtime_error &E) {
+    std::cerr << TestCaseName
+              << " failed: unexpected runtime_error exception: " << E.what()
+              << std::endl;
+    return 1;
+  } catch (...) {
+    std::cerr << TestCaseName << " failed: something unexpected has been caught"
+              << std::endl;
+    return 1;
+  }
+}
+
+template <int... ReqdWGSizes>
+int runPositiveTest(queue &Q, range<sizeof...(ReqdWGSizes)> GlobalRange) {
+  std::string TestCaseName =
+      "Test case ReqdWGSizeVariadic " +
+      GetRangeStr(range<sizeof...(ReqdWGSizes)>{ReqdWGSizes...}) +
+      " (Positive without local range)";
+  try {
+    submitTest<TestKind::PositiveWithoutLocal, ReqdWGSizes...>(Q, GlobalRange);
+    Q.wait_and_throw();
+    return 0;
+  } catch (nd_range_error &E) {
```

Comment:
The submitted kernel uses 'range', not 'nd_range'. So, why expecting nd_range_error?
Also, nd_range_error is deprecated and should not be used in new tests, IMO.

Reply:
Good point. I suspect it was based on the knowledge that if we throw for an invalid reqd_work_group_size + local work size combination it would always be an `nd_range_error`, primarily because it should only happen for `nd_range` enqueues. However, I agree that we may as well move away from it now in the test, so I have changed the test to either catch a `sycl::exception` or whatever else is thrown. Note though that the runtime will still produce it for the negative case until we change in in the runtime, which we should be doing soon with some other related changes.